File: src/main/java/org/mitallast/queue/crdt/routing/RoutingReplica.java
package org.mitallast.queue.crdt.routing;

import org.mitallast.queue.common.codec.Codec;
import org.mitallast.queue.common.codec.Message;
import org.mitallast.queue.transport.DiscoveryNode;

public class RoutingReplica implements Message {
    public static final Codec<RoutingReplica> codec = Codec.of(
        RoutingReplica::new,
        RoutingReplica::id,
        RoutingReplica::member,
        RoutingReplica::state,
        Codec.longCodec,
        DiscoveryNode.codec,
        Codec.enumCodec(State.class)
    );

    public enum State {
        OPENED, CLOSED;
    }

    private final long id;
    private final DiscoveryNode member;
    private final State state;

    public RoutingReplica(long id, DiscoveryNode member) {
        this(id, member, State.OPENED);
    }

    public RoutingReplica(long id, DiscoveryNode member, State state) {
        this.id = id;
        this.member = member;
        this.state = state;
    }

    public long id() {
        return id;
    }

    public DiscoveryNode member() {
        return member;
    }

    public State state() {
        return state;
    }

    public boolean isOpened() {
        return state == State.OPENED;
    }

    public boolean isClosed() {
        return state == State.CLOSED;
    }

    public RoutingReplica close() {
        return new RoutingReplica(id, member, State.CLOSED);
    }

    @Override
    public String toString() {
        return "RoutingReplica{id=" + id + ", " + member + ", state=" + state + '}';
    }
}

File: src/main/java/org/mitallast/queue/crdt/routing/RoutingReplica.kt
package org.mitallast.queue.crdt.routing

import org.mitallast.queue.common.codec.Codec
import org.mitallast.queue.common.codec.Message
import org.mitallast.queue.transport.DiscoveryNode

class RoutingReplica @JvmOverloads constructor(
    val id: Long,
    val member: DiscoveryNode,
    val state: State = State.OPENED) : Message {

    enum class State {
        OPENED, CLOSED
    }

    val isOpened: Boolean
        get() = state == State.OPENED

    val isClosed: Boolean
        get() = state == State.CLOSED

    fun close(): RoutingReplica {
        return RoutingReplica(id, member, State.CLOSED)
    }

    companion object {
        val codec = Codec.of(
            ::RoutingReplica,
            RoutingReplica::id,
            RoutingReplica::member,
            RoutingReplica::state,
            Codec.longCodec(),
            DiscoveryNode.codec,
            Codec.enumCodec(State::class.java)
        )
    }
}
